Articles/Blog/WhenIsBetterToWriteAutomatedTests.md

```diff
@@ -0,0 +1,133 @@
+
+## Criteria for when it is better to not add automated tests
+
+I have come to realize that it often does not pay off to write automated tests for a piece of software when the following five things are true about that software:
+
+* **The damage done by a defect is minor.**
+* **The majority of defects will be obvious.**
+* **Fixing the majority of defects will be easy.**
+* **Manually testing the software after a change is easy.**
+* **Writing automated tests is hard or the tests will be hard to maintain.**
+
+If all five of the above criteria are satisfied, it is often better not to write automated tests for a piece of software.
+The time spent writing, maintaining, and running the automated tests does not have enough benefit to outweigh the cost.
+
+For example, some bash scripts used locally that load some modules and then configure, build, and run tests for some software may not be worth writing an automated test suite for because:
+
+* The damage done by not having everything complete correctly will be minor because you can just run it again.
+* A failure will be obvious because the software will not even build and no tests will be run.
+* The scripts do pretty basic stuff and will be easy to fix if they fail.
+* It would be difficult to write mock up the various commands run by the script.
+
+Therefore, it is often not worth the investment to write automated tests for locally run scripts of this type.
+
+## Example where automated tests should have been added
+
+However, if such scripts are being used to automate some important process like deploying the software to users where not performing the task correctly would cause non-trivial harm, then one should write some automated tests to protect the critical functionality of the scripts.
+Or, if other people are running the scripts and it will not be obvious to them that a failure has occurred or they will not know how to fix it quickly, then one should likely write some automated tests for the scripts.
+
+For example, there was a bash test-driver script in a CSE project where the script (run as a cron job) failed to detect that the code's tests were failing and instead sent out emails that everything was passing.
+As a result, the code was broken for weeks with no one noticing.
+A release of the software went out (to hundreds of internal users) and it was the users who noticed the new defects.
+This wasted user's time, may have resulted in incorrect results, and damaged the reputation of the project releasing the software.
+The lesson is that some *scripts* may actually need to be considered *software* in their own right and need to have strong automated testing just like any piece of non-trivial software.
+(Just become some piece of software is written in bash or python does not mean it can be dismissed as "scripts" and avoid any automated testing.)
```

Comment:
perhaps a comment about proper selection of language to facilitate testing here instead of this type of overhead comment about 2 languages. Python has solid testing infrastructure in place for instance while bash has nothing, ruby has excellent testing in place, while fortran is just beginning to add testing.

Reply:
@prwolfe said:

> perhaps a comment about proper selection of language to facilitate testing here instead of this type of overhead comment about 2 languages

The issue of language is largely an excuse.  For example, there is no unit test harnesses for CMake code that I now about but you can see that with a little work you can write a simple one:

* https://github.com/TriBITSPub/TriBITS/blob/master/tribits/core/utils/UnitTestHelpers.cmake

An example of its usage is:

* https://github.com/TriBITSPub/TriBITS/blob/master/test/core/TestingFunctionMacro_UnitTests.cmake 

which is run with `ctest -P TestingFunctionMacro_UnitTests.cmake` in:

* https://github.com/TriBITSPub/TriBITS/blob/master/test/core/CMakeLists.txt#L44

Yes, having access to a good unit test harness for your language helps but the larger issue is the nature of what is being tested, not the language.  For example, if you rewrote the bash script:

* https://github.com/trilinos/Trilinos/blob/develop/cmake/std/PullRequestLinuxDriver-Test.sh

in Python, would that make it massively easier to test?  Yes, it would make it a little easier but the hard stuff is that it calls git commands and it runs builds and tests and has to respond to all types of failures conditions and report those correctly when they occur for all use cases.  Can you test code that does stuff like git commands, running builds and tests, etc?  Yes you can.  For example:

* https://github.com/TriBITSPub/TriBITS/blob/master/test/ci_support/CheckinTest_UnitTests.py
* https://github.com/TriBITSPub/TriBITS/blob/master/test/ctest_driver/TribitsExampleProject/CMakeLists.txt

I don't want to give people the language excuse for not writing automated tests.  You can write automated tests in any Turing-complete language.

Reply:
I agree that the choice of language is not an excuse, but correct selection can make everyones life easier.

Reply:
@prwolfe, hopefully this is addressed in https://github.com/betterscientificsoftware/betterscientificsoftware.github.io/pull/377/commits/00243f238f4e0a12535e805d9fe6dcd7fd2f30e9?